riscv-page-tables/src/page_table.rs

```diff
@@ -1207,6 +1269,137 @@ impl<T: PagingMode> GuestStagePageTable<T> {
+    /// smaller than the requested page size. Also checks the pages are physically contiguous and
+    /// suitably aligned to the requested page size. It's important that all the pages must be in
+    /// the same state, either "Blocked" and uniquely owned by the TVM, or "BlockedShared" and not
+    /// owned by the TVM. Last thing to check is that each page must have been invalidated with a
+    /// TLB version that's older than the current one.
+    pub fn promote_page(
+        &self,
+        vaddr: PageAddr<T::MappedAddressSpace>,
+        requested_page_size: PageSize,
+        expected_page_size: PageSize,
+        page_tracker: PageTracker,
+        owner: PageOwnerId,
+        tlb_version: TlbVersion,
+    ) -> Result<(SupervisorPageAddr, PageSize, Page<InternalClean>)> {
+        let num_pages = PageSize::num_4k_pages(requested_page_size as u64);
+        let end = vaddr
+            .checked_add_pages(num_pages)
+            .ok_or(Error::AddressOverflow)?;
+        let mut inner = self.inner.lock();
+        let mut prev_page_state = None;
+        let mut prev_paddr: Option<SupervisorPageAddr> = None;
+        let mut va = vaddr;
+        while va < end {
+            let pte = inner.get_invalidated_leaf(va)?;
+            let paddr = pte.page_addr();
+            let page_size = pte.level().leaf_page_size();
+
+            // Check the page matches the expected page size (one size smaller than the
+            // requested page size)
+            if page_size != expected_page_size {
+                return Err(Error::InvalidPageSize);
+            }
+
+            // Perform some checks on the physical address
+            if let Some(addr) = &prev_paddr {
+                // Check the current page is contiguous with the previous page
+                if addr
+                    .checked_add_pages_with_size(1, expected_page_size)
+                    .ok_or(Error::AddressOverflow)?
+                    != paddr
+                {
+                    return Err(Error::PageNotContiguous);
+                }
+            } else {
+                // This is the first page, we must ensure it's suitably aligned with the requested
+                // page size.
+                if !paddr.is_aligned(requested_page_size) {
+                    return Err(Error::AddressMisaligned(paddr.bits()));
+                }
+            }
+            prev_paddr = Some(paddr);
+
+            // Perform some checks on the page state
+            let page_state = page_tracker
+                .page_state(paddr, page_size)
+                .ok_or(Error::InvalidPageState)?;
+            if !page_tracker.is_promotable_page(
+                paddr,
+                page_size,
+                owner,
+                MemType::Ram,
+                prev_page_state,
+                tlb_version,
+            ) {
+                return Err(Error::PageNotPromotable);
+            }
```

Comment:
to be consistent with the other `PageTable` methods we should do this `PageTracker` checking in a closure provided by the caller. and the caller can determine whether the page to be promoted lies within a confidential or non-confidential memory region, so they need only check whether it's blocked at an older tlb version (i.e. no need to check that the state of each page is the same).

Reply:
Ah ok I understand. The reason I couldn't use the closure design was because I wanted to check every page state was the same, but if we rely on memory region for that, then that should solve the problem.